crates/ruff_linter/src/rules/ruff/rules/non_string_literal_as_assert_message.rs

```diff
@@ -0,0 +1,59 @@
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{Expr, StmtAssert};
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for uses of non-string literal as assert message.
+///
+/// ## Why is this bad?
+/// Non-string literal in assert message does not provide any useful
+/// information and is likely an unitentional use of `assert_equal(expr, expr)`
+/// from other languages.
+///
+/// ## Example
+/// ```python
+/// fruits = ["apples", "plums", "pears"]
+/// fruits.filter(lambda fruit: fruit.startwith("p"))
+/// assert len(fruits), 2  # True unless the list is empty
+/// ```
+///
+/// Use instead:
+/// ```python
+/// fruits = ["apples", "plums", "pears"]
+/// fruits.filter(lambda fruit: fruit.startwith("p"))
+/// assert len(fruits) == 2
+/// ```
+#[violation]
+pub struct NonStringLiteralAsAssertMessage;
```

Comment:
It's a little confusing that `NonStringLiteral` can be parsed as either:

- non-string literal, or
- non - (string-literal)

the former means "a literal that is not a string literal" the latter would include like... arbitrary expressions.

Can we rename the rule to `literal-besides-string-as-assert-message` or `literal-other-than-string-as-assert-message`?

Reply:
literal-other-than-string-as-assert-message sounds ok

Reply:
sorry for joining the bikeshedding: How about `invalid-assert-message-literal-argument`

It's less specific because it doesn't mention string literals but I think that's a benefit if we decide to allow byte literals.

Reply:
I accept this bike shed color begrudgingly since "invalid" seems a bit strong - but let's keep it for now, changing the human-readable names can be done at any time since it's not a breaking change.